### ISSUE TITLE
Trains and Boats are auto reversed when path is too long and there is a single vehicule.

### ISSUE THREAD

**Issue by Muchaszewski** [TommyC81/TPF2-LineManager#13]
**Issue**

This is especially visible when playing starting from 1850 on Hard. You can build a single train or single boat (more boats are unnecessary) but the path is usually long due to low speed of the boats.

The issue is that "reverse" is pressed on the vehicle if nothing is wrong with the path or vehicle!

**Reproduction steps**
Starting in 1850, build a long train or long boat that takes >1 year to move cargo once.

![image](https://user-images.githubusercontent.com/5092066/235006757-0fbadddf-c7d0-403d-be4a-e5c911c10f00.png)

**Current result**
If Line Manager is enabled (other options don't matter) your vehicles will be reversed after a few dozen seconds.

**Expected result**
Nothing happens because the path is correct, there is cargo, the path is accessible, etc. It just takes over 1 year.

I hope you can reproduce this issue because this is very annoying. 

**Additional info**
 - Line Manger enabled - Only this option affects this behavior (fixes it)
 
 - Congestion control enabled - Doesn't affect this behavior (still present)
 - Automatically reverse trains with no path - Doesn't affect this behavior
 - Cargo Water/Rail - Doesn't affect this behavior
 - Adding [M] at the end - Doesn't affect this behavior

**Comment by TommyC81** [comment]
Please send the output from console (you can check stdout.txt), and ideally also prepare a clean game (only LineManager) that demonstrates the problem.

At face value, the game is trying to add vehicles to meet a minimum frequency of 12 minutes. You either don't have enough money, or no depot available, causing it to fail. The vehicles turning around is seemingly the attempt to find a depot.

The solution for you is to either disable LineManager
initially, or set these extremely long routes to manual management ("[M]") - or is this not working? - until you have enough money to allow them to be effectively managed.

Either way, please see the top paragraph to provide more data.

**Comment by TommyC81** [comment]
An obvious question: Are you sure you have a shipyard available?

**Comment by Muchaszewski** [comment]
I will send everything as a requested later today.

Yes I'm absolutely sure that shipyard is available. This is very basic setup that works if the line manager is disabled. The caveat is that as you said.

A - I might have not enough money for new ship.
B - The path longer than 12 minutes

**Comment by Muchaszewski** [comment]
1. Adding [M] at the end - Doesn't affect this behavior (still present)
I cannot reproduce the one case where [M] at line name would not prevent this behavior. My line names are usually like this:
`[C] Oxford-Cambridge Wheat #1` and after adding [M] -> `[C] Oxford-Cambridge Wheat #1 [M]` 

but! From the logs I get this, shouldn't it not be in the logs if there is [M] at the end?
```
[LineManager][19:57:31][WARN] Unable to add vehicle to line '[C] Line 2 [M]' - Insufficient cash?
```

2. Cargo Water/Rail - Doesn't affect this behavior
![image](https://user-images.githubusercontent.com/5092066/235218819-2c9e250c-fc51-4aa4-8e0d-52c298424daf.png)
![image](https://user-images.githubusercontent.com/5092066/235218962-5d5cf191-aa8c-49e0-ab43-1ecb100e999f.png)
Also disabling Cargo-Water works here, but didn't at my previous save which I cannot reproduce right now. :( Although when I was checking out the train, with Cargo-Water disabled Line 1 ship lost its cargo

Description of this save.
Line 1 - is broken in the way I described. I'm attaching a save file, stdout right after the issue happens. Other lines function correctly unfortunately but wanted to showcase the same behavior for rail and water. Also stdout2.txt contains about 20 minutes of logs where ship on line 1 never arrives at it's destination.

[save.zip](https://github.com/TommyC81/TPF2-LineManager/files/11356056/save.zip)
[stdout.txt](https://github.com/TommyC81/TPF2-LineManager/files/11356071/stdout.txt)
[stdout2.txt](https://github.com/TommyC81/TPF2-LineManager/files/11356206/stdout.txt)

The last bit is a video. I try to showcase the issue. In the video (as it is in the savegame) the stations are reversed, but that should not matter for reproducibility. I click nothing, and the ship just reverses on it's own! :)
https://user-images.githubusercontent.com/5092066/235222007-dadbfca5-f77b-4ee3-be8e-b8689548b85f.mp4

I hope you can find what causes that!
Thank you for your effort and this great mod! I'm loving it when it works :)

**Comment by TommyC81** [comment]
Thanks for this, I will take a closer look in the coming day.


I notice one error: You cannot assign multiple rules to a line. If you do, only the first one will be used ie if you have [C] followed by [M], then CARGO rules will be applied (the line will be managed,regardless of whether automatic setting is enabled or not).


Note: [M] is not a real rule per se, it simply tells the mod that you're managing that line manually by yourself and the mod should ignore it. However, if you've added a [C] rule prior, then you've explicitly asked the mod to manage the line as per [C] rules.


What you should do, probably, is simply ONLY disable automatic CARGO / WATER line management in the settings, and you're done. No need to add any
specific rules for any CARGO / WATER lines as they won't be managed. If the automatic setting is disabled, and you want a specific line to be managed anyway, only then add a rule to the line name ie [C].


The only other occasion you'd manually assign a rule is of you want another than the default role applied, for instance the rate rule instead, example: [R:200].


In summary: it seems you've disabled automatic management, but then specifically told the mod to manage the line anyway (the second rule in the same line name, ie [M], is ignored).


I hope this makes sense (from my phone).

**Comment by Muchaszewski** [comment]
Oh so [C] is cargo rule. That might explain things. I just kept this naming convention from my other games to categorize what route does what. 
I used something like C for Cargo, D for Delivery, IC for Intercity travel and so on. I will stop using square brackets then for my stuff if I want to use this mod 😅

I also for some reason though that you can only add rules at the end of the name. But just read readme again and it says clearly that you can add them anywhere. 

So all of this makes sense now. Some of the stuff it's just me not reading instructions right. But the main issue remains! Why does the ship reverses (as said before this also happened to train). 

Thank you so much!

**Comment by TommyC81** [comment]
Thanks for providing the files.

The following can be observed:

- There's no cash, so any action that LineManager takes will fail. When something fails, it will try to find a new depot again the next run (to rule out possible issues) - when a vehicle is sent to a depot, it empties the vehicle (this is hard-coded behaviour in Transport Fever 2, however, depots are cached by LineManager, reducing the occurrence of this as time goes), thus you only see repeatedly emptying of vehicles.
- When "Insufficient cash" is the most likely cause, a warning message will be in the console: LineManager][12:16:26][WARN] Unable to add vehicle to line 'Line 1' - Insufficient cash?
- If there's no depot accessible for a vehicle when trying to find a depot to add more vehicles to that line, then locating a depot will fail. When this is the case, a warning message will be in the console: [LineManager][12:16:26][WARN] Unable to add vehicle to line 'Line 3' - Either no available depot, or not possible to find a depot on this update.
- The [M] rule (and all other settings) work as expected, and this can also be seen in the console (if you enable "Show extended line info": 
![image](https://user-images.githubusercontent.com/1904092/235292736-8b05f5ce-4a8e-46db-a626-e0281231205c.png)

Anyway, to get the behaviour you want, you should do the following:

1. **(Please see point 1 in 'Quick tips for best results' in the manual)**
2. Disable automatic line management for all/most types of lines (except perhaps road vehicles as they are individually cheaper), until you have enough cash so LineManager can work properly without being cash limited.
3. In lieu of automatic management for whole types of lines, add rules where you really want them i.e. [C], [R:200], [P]. When you manually add rules, they will always be processed (unless LineManager is disabled).
4. For Line 3 (train route) to work properly with LineManager, you must connect the depot so it is accessible from all directions. In the current setup, the depot can only be reached from the direction of station "Fort Lauderdale", so a train that has passed the depot will never be able to locate it as there is no accessible path. If you let the game run for a bit, you will notice that you will get an "Insufficient cash" warning message when the train is between station 'Fort Lauderdale' and the depot, as the depot then is accessible and can be used, but there's not sufficient cash to buy a vehicle. (Locating the depot will also _empty the vehicle_, as that is Transport Fever 2 behavior when a vehicle is sent to a depot - however, again, depots are cached, reducing the occurrence of this as time goes.)
![image](https://user-images.githubusercontent.com/1904092/235293658-24d9e486-1ab5-4b96-864a-bab756f8b962.png)


All-in-all, everything seems to be working as expected. At least I cannot see any obvious issue (that is not already a known limitation). I hope this helps.

Final note: If you check further in the manual, you'll see in the 'Bounty hunt' section that there are a few items that are of specific interest to improve/resolve - if you have any advice on how to achieve it, it would be much appreciated :-)

**Comment by TommyC81** [comment]
I've updated the vehicle addition logic in this commit: c1488f0bdd94d41e22aafe6ce18447e8c650c9bf

This should significantly reduce the "reversing" you've observed, and should in general make things behave a bit more predictable and "as expected". It is live on the workshop since yesterday.

**Comment by TommyC81** [comment]
@Muchaszewski I just want to confirm that the above, as well as the recent code update, has perhaps resolved this matter for now?

**Comment by Muchaszewski** [comment]
Sounds great! I had not yet a chance to test that, as I am in vacations but will check it out next week and report back!

**Comment by TommyC81** [comment]
Nice one, enjoy your vacation!

**Comment by Muchaszewski** [comment]
I just tried it and would say it's 99% better! Not perfect as I saw it reverse once (starting from the save file after 5-10 seconds into the save on Line 1), but I think you got it!

I gave it a try on a new game and even though it failed once or twice. But it was not an issue anymore. I didn't have to go back to each line and babysit them. They didn't lose cargo, and went through the whole way uninterrupted!

Great work! This is the best mod for TPF2! 

Now I need to find something that will auto-name my lines :D
